File: src/cpp/model_manager.cpp
#include <cstdlib>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <map>
#include <optional>
#include <string>
#include <vector>
#include <algorithm>
#include <cctype>
#include <unordered_set>

#ifdef _WIN32
#define WIN32_LEAN_AND_MEAN
#define NOMINMAX
#include <windows.h>
#elif defined(__APPLE__)
#include <mach-o/dyld.h>
#include <climits>
#else
#include <climits>
#endif

#include <spdlog/spdlog.h>
#include "json.hpp"
#include "model_manager.hpp"

using json = nlohmann::json;
namespace fs = std::filesystem;

namespace piper {

// ---------------------------------------------------------------------------
// Embedded piper-plus voice catalog
// ---------------------------------------------------------------------------

static const char* PIPER_PLUS_CATALOG_JSON = R"JSON(
{
    "ja_JP-tsukuyomi-chan-medium": {
        "key": "ja_JP-tsukuyomi-chan-medium",
        "name": "tsukuyomi-chan",
        "language": {
            "code": "ja_JP", "family": "ja",
            "name_native": "日本語", "name_english": "Japanese"
        },
        "quality": "medium",
        "num_speakers": 1,
        "speaker_id_map": {},
        "source": "piper-plus",
        "repo": "ayousanz/piper-plus-tsukuyomi-chan",
        "files": {
            "tsukuyomi-chan-6lang-fp16.onnx": {
                "size_bytes": 39216913,
                "md5_digest": ""
            },
            "config.json": {
                "size_bytes": 8568,
                "md5_digest": ""
            }
        },
        "aliases": ["tsukuyomi", "tsukuyomi-chan", "ja-tsukuyomi"],
        "description": "Tsukuyomi-chan 6-language TTS model fine-tuned from multilingual base (FP16)"
    },
    "ja_JP-css10-6lang-medium": {
        "key": "ja_JP-css10-6lang-medium",
        "name": "css10-6lang",
        "language": {
            "code": "ja_JP", "family": "ja",
            "name_native": "日本語", "name_english": "Japanese"
        },
        "quality": "medium",
        "num_speakers": 1,
        "speaker_id_map": {},
        "source": "piper-plus",
        "repo": "ayousanz/piper-plus-css10-ja-6lang",
        "files": {
            "css10-ja-6lang-fp16.onnx": {
                "size_bytes": 39414515,
                "md5_digest": ""
            },
            "config.json": {
                "size_bytes": 8966,
                "md5_digest": ""
            }
        },
        "aliases": ["css10", "css10-6lang", "css10-ja", "ja-css10"],
        "description": "CSS10 Japanese 6-language TTS model fine-tuned from multilingual base (FP16, 6841 utterances)"
    }
}
)JSON";

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

// Get the directory containing the running executable
static fs::path getExeDir() {
#ifdef _WIN32
    wchar_t wbuf[MAX_PATH] = {0};
    DWORD len = GetModuleFileNameW(nullptr, wbuf, MAX_PATH);
    if (len > 0 && len < MAX_PATH) {
        return fs::path(wbuf).parent_path();
    }
#elif defined(__APPLE__)
    char buf[PATH_MAX] = {0};
    uint32_t size = sizeof(buf);
    if (_NSGetExecutablePath(buf, &size) == 0) {
        return fs::path(buf).parent_path();
    }
#else
    try {
        return fs::canonical("/proc/self/exe").parent_path();
    } catch (...) {
        // fall through
    }
#endif
    return fs::current_path();
}

// Parse a single voice entry from JSON into VoiceInfo
static VoiceInfo parseVoiceEntry(const std::string& key,
                                 const json& entry,
                                 const std::string& defaultSource) {
    VoiceInfo vi;
    vi.key = key;
    vi.name = entry.value("name", "");
    vi.quality = entry.value("quality", "");
    vi.numSpeakers = entry.value("num_speakers", 1);
    vi.source = entry.value("source", defaultSource);

    // HuggingFace repo (piper-plus voices)
    vi.repoId = entry.value("repo", "");

    // Language block
    if (entry.contains("language") && entry["language"].is_object()) {
        const auto& lang = entry["language"];
        vi.languageCode = lang.value("code", "");
        vi.languageFamily = lang.value("family", "");
        vi.languageNameNative = lang.value("name_native", "");
        vi.languageNameEnglish = lang.value("name_english", "");
    }

    // Files
    if (entry.contains("files") && entry["files"].is_object()) {
        for (auto& [filePath, fileInfo] : entry["files"].items()) {
            VoiceFileInfo vfi;
            vfi.relativePath = filePath;
            vfi.sizeBytes = fileInfo.value("size_bytes", (size_t)0);
            vfi.md5Digest = fileInfo.value("md5_digest", "");
            vi.files.push_back(std::move(vfi));
        }
    }

    // Aliases
    if (entry.contains("aliases") && entry["aliases"].is_array()) {
        for (const auto& alias : entry["aliases"]) {
            if (alias.is_string()) {
                vi.aliases.push_back(alias.get<std::string>());
            }
        }
    }

    return vi;
}

// Try to locate the upstream voices.json file near the executable
static std::optional<fs::path> findUpstreamVoicesJson() {
    fs::path exeDir = getExeDir();

    // Search paths relative to executable
    std::vector<fs::path> candidates = {
        exeDir / "voices.json",
        exeDir / ".." / "share" / "voices.json",
        exeDir / ".." / "share" / "piper" / "voices.json",
    };

    for (const auto& p : candidates) {
        std::error_code ec;
        if (fs::exists(p, ec)) {
            auto canon = fs::canonical(p, ec);
            if (!ec && fs::exists(canon)) {
                return canon;
            }
        }
    }

    return std::nullopt;
}

// Shell-safe for URLs: allowlist approach.
// Only allow alphanumerics, hyphens, underscores, dots, forward slashes,
// colons, and percent (for URL-encoded characters).
// Explicitly rejects shell metacharacters: ' $ ` ( ) ; | & < > ~ # ! { } etc.
static bool isSafeForShell(const std::string& s) {
    for (char c : s) {
        if (!std::isalnum(static_cast<unsigned char>(c)) &&
            c != '-' && c != '_' && c != '.' && c != '/' &&
            c != ':' && c != '%') {
            return false;
        }
    }
    return !s.empty();
}

// Shell-safe for file paths: allows backslashes for Windows path separators.
// Explicitly rejects shell metacharacters: ' $ ` ( ) ; | & < > ~ # ! { } etc.
static bool isSafeForShellPath(const std::string& s) {
    for (char c : s) {
        if (!std::isalnum(static_cast<unsigned char>(c)) &&
            c != '-' && c != '_' && c != '.' && c != '/' &&
            c != '\\' && c != ':') {
            return false;
        }
    }
    return !s.empty();
}

// Validate that a voice key contains no path traversal characters.
// Rejects "..", "/", and "\" to prevent directory escape.
static bool isSafeVoiceKey(const std::string& key) {
    if (key.empty()) return false;
    if (key.find("..") != std::string::npos) return false;
    if (key.find('/') != std::string::npos) return false;
    if (key.find('\\') != std::string::npos) return false;
    return true;
}

// Validate that a repoId contains only safe characters (alphanumerics, hyphens,
// underscores, dots, and a single forward slash separating owner/repo).
static bool isSafeRepoId(const std::string& repoId) {
    if (repoId.empty()) return false;
    int slashCount = 0;
    for (char c : repoId) {
        if (c == '/') {
            ++slashCount;
            if (slashCount > 1) return false;  // only one slash allowed
        } else if (!std::isalnum(static_cast<unsigned char>(c)) &&
                   c != '-' && c != '_' && c != '.') {
            return false;
        }
    }
    // Must have exactly one slash (owner/repo format)
    return slashCount == 1;
}

// Download a single file using system() with curl/wget/PowerShell
static bool downloadFile(const std::string& url,
                         const fs::path& outputPath) {
    // Validate url and path for shell safety
    if (!isSafeForShell(url)) {
        spdlog::error("URL contains unsafe characters: {}", url);
        return false;
    }

    std::string outStr = outputPath.string();
    // Allow isSafeForShellPath characters plus spaces (paths are quoted in commands).
    if (!isSafeForShellPath(outStr)) {
        // isSafeForShellPath rejects spaces; check if spaces are the only extras
        bool safeWithSpaces = !outStr.empty();
        for (char c : outStr) {
            if (c != ' ' && !std::isalnum(static_cast<unsigned char>(c)) &&
                c != '-' && c != '_' && c != '.' && c != '/' &&
                c != '\\' && c != ':') {
                safeWithSpaces = false;
                break;
            }
        }
        if (!safeWithSpaces) {
            spdlog::error("Output path contains unsafe characters: {}", outStr);
            return false;
        }
    }

    std::string cmd;

#ifdef _WIN32
    cmd = "powershell -NoProfile -Command \"Invoke-WebRequest -Uri '"
        + url + "' -OutFile '" + outStr + "'\"";
#else
    // Prefer curl, fall back to wget
    if (std::system("which curl > /dev/null 2>&1") == 0) {
        cmd = "curl -L -# -o \"" + outStr + "\" \"" + url + "\"";
    } else if (std::system("which wget > /dev/null 2>&1") == 0) {
        cmd = "wget -O \"" + outStr + "\" \"" + url + "\"";
    } else {
        spdlog::error("Neither curl nor wget is available for downloading");
        return false;
    }
#endif

    spdlog::info("Downloading {} ...", url);
    int rc = std::system(cmd.c_str());
    return rc == 0;
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

std::filesystem::path getDefaultModelDir() {
    // Environment variable takes precedence
    const char* envDir = std::getenv("PIPER_MODEL_DIR");
    if (envDir && envDir[0] != '\0') {
        return fs::path(envDir);
    }

#ifdef _WIN32
    const char* appData = std::getenv("APPDATA");
    if (appData) {
        return fs::path(appData) / "piper" / "models";
    }
    return fs::path("models");
#elif defined(__APPLE__)
    const char* home = std::getenv("HOME");
    if (home) {
        return fs::path(home) / "Library" / "Application Support"
               / "piper" / "models";
    }
    return fs::path("models");
#else
    const char* xdgData = std::getenv("XDG_DATA_HOME");
    if (xdgData) {
        return fs::path(xdgData) / "piper" / "models";
    }
    const char* home = std::getenv("HOME");
    if (home) {
        return fs::path(home) / ".local" / "share" / "piper" / "models";
    }
    return fs::path("models");
#endif
}

std::vector<VoiceInfo> loadVoiceCatalog() {
    std::vector<VoiceInfo> catalog;

    // 1. Parse embedded piper-plus catalog
    try {
        json piperPlusCatalog = json::parse(PIPER_PLUS_CATALOG_JSON);
        for (auto& [key, entry] : piperPlusCatalog.items()) {
            catalog.push_back(parseVoiceEntry(key, entry, "piper-plus"));
        }
        spdlog::debug("Loaded {} piper-plus voices from embedded catalog",
                       catalog.size());
    } catch (const json::exception& e) {
        spdlog::error("Failed to parse embedded piper-plus catalog: {}",
                       e.what());
    }

    // 2. Try to load upstream voices.json
    auto upstreamPath = findUpstreamVoicesJson();
    if (upstreamPath) {
        try {
            std::ifstream ifs(upstreamPath.value());
            json upstreamCatalog = json::parse(ifs);

            // Build a set of existing keys for O(1) duplicate lookup
            std::unordered_set<std::string> existingKeys;
            existingKeys.reserve(catalog.size());
            for (const auto& existing : catalog) {
                existingKeys.insert(existing.key);
            }

            size_t count = 0;
            for (auto& [key, entry] : upstreamCatalog.items()) {
                // Skip if already present (piper-plus overrides upstream)
                if (existingKeys.count(key) == 0) {
                    // Validate repoId from external JSON to reject malicious entries
                    std::string repo = entry.value("repo", "");
                    if (!repo.empty() && !isSafeRepoId(repo)) {
                        spdlog::warn("Skipping upstream entry '{}': "
                                     "unsafe repoId '{}'", key, repo);
                        continue;
                    }
                    catalog.push_back(parseVoiceEntry(key, entry, "piper"));
                    existingKeys.insert(key);
                    ++count;
                }
            }
            spdlog::debug("Loaded {} upstream voices from {}",
                           count, upstreamPath.value().string());
        } catch (const std::exception& e) {
            spdlog::warn("Failed to load upstream voices.json ({}): {}",
                          upstreamPath.value().string(), e.what());
        }
    } else {
        spdlog::debug("No upstream voices.json found; "
                       "using piper-plus catalog only");
    }

    // Sort by language code, then by key
    std::sort(catalog.begin(), catalog.end(),
              [](const VoiceInfo& a, const VoiceInfo& b) {
                  if (a.languageCode != b.languageCode)
                      return a.languageCode < b.languageCode;
                  return a.key < b.key;
              });

    return catalog;
}

std::optional<VoiceInfo> findVoice(const std::string& nameOrAlias) {
    auto catalog = loadVoiceCatalog();

    // 1. Exact key match
    for (const auto& voice : catalog) {
        if (voice.key == nameOrAlias) {
            return voice;
        }
    }

    // 2. Alias match (no partial matching -- aliases provide sufficient coverage)
    for (const auto& voice : catalog) {
        for (const auto& alias : voice.aliases) {
            if (alias == nameOrAlias) {
                return voice;
            }
        }
    }

    return std::nullopt;
}

void listModels(const std::string& languageFilter) {
    auto catalog = loadVoiceCatalog();

    if (catalog.empty()) {
        std::cerr << "No voice models found." << std::endl;
        return;
    }

    // Filter by language if specified
    std::vector<VoiceInfo> filtered;
    if (languageFilter.empty()) {
        filtered = catalog;
    } else {
        for (const auto& voice : catalog) {
            if (voice.languageFamily == languageFilter ||
                voice.languageCode == languageFilter) {
                filtered.push_back(voice);
            }
        }
    }

    if (filtered.empty()) {
        std::cerr << "No voice models found for language: "
                  << languageFilter << std::endl;
        return;
    }

    std::cerr << std::endl;
    std::cerr << "Available voice models:" << std::endl;

    // Group by language code
    std::string currentLang;
    for (const auto& voice : filtered) {
        if (voice.languageCode != currentLang) {
            currentLang = voice.languageCode;
            std::cerr << std::endl;
            std::cerr << "  " << voice.languageNameEnglish;
            if (!voice.languageNameNative.empty() &&
                voice.languageNameNative != voice.languageNameEnglish) {
                std::cerr << " (" << voice.languageNameNative << ")";
            }
            std::cerr << " [" << voice.languageCode << "]:" << std::endl;
        }

        // Format: key  [source]  N speaker(s)  quality  (aliases)
        std::cerr << "    " << voice.key;

        // Pad to 40 chars for alignment
        int padLen = 40 - static_cast<int>(voice.key.size());
        if (padLen > 0) {
            std::cerr << std::string(padLen, ' ');
        } else {
            std::cerr << "  ";
        }

        std::cerr << "[" << voice.source << "]  ";
        std::cerr << voice.numSpeakers << " speaker"
                  << (voice.numSpeakers != 1 ? "s" : "") << "   ";
        std::cerr << voice.quality;

        if (!voice.aliases.empty()) {
            std::cerr << "   (";
            for (size_t i = 0; i < voice.aliases.size(); ++i) {
                if (i > 0) std::cerr << ", ";
                std::cerr << voice.aliases[i];
            }
            std::cerr << ")";
        }
        std::cerr << std::endl;
    }

    std::cerr << std::endl;
    std::cerr << "Use --download-model <name> to download a model."
              << std::endl;
    std::cerr << std::endl;
}

bool downloadModel(const std::string& modelName,
                   const fs::path& modelDir) {
    // Look up the voice in the catalog
    auto maybeVoice = findVoice(modelName);
    if (!maybeVoice) {
        spdlog::error("Model '{}' not found. "
                       "Use --list-models to see available models.",
                       modelName);
        return false;
    }

    const VoiceInfo& voice = maybeVoice.value();
    spdlog::info("Downloading model: {} ({})", voice.key, voice.source);

    // Reject voice.key with path traversal characters (fix: path traversal prevention)
    if (!isSafeVoiceKey(voice.key)) {
        spdlog::error("Voice key '{}' contains unsafe path characters", voice.key);
        return false;
    }

    // Flat directory layout matching Python: files go directly into modelDir/
    fs::path targetDir = modelDir;
    std::error_code ec;
    fs::create_directories(targetDir, ec);
    if (ec) {
        spdlog::error("Failed to create directory {}: {}",
                       targetDir.string(), ec.message());
        return false;
    }

    // Build base URL depending on source
    std::string baseUrl;
    if (voice.source == "piper-plus") {
        // https://huggingface.co/{repo}/resolve/main/{filename}
        baseUrl = "https://huggingface.co/" + voice.repoId + "/resolve/main/";
    } else {
        // Upstream piper:
        // https://huggingface.co/rhasspy/piper-voices/resolve/v1.0.0/{file_path}
        baseUrl = "https://huggingface.co/rhasspy/piper-voices/resolve/v1.0.0/";
    }

    // Download each file
    bool allOk = true;
    for (const auto& file : voice.files) {
        std::string url = baseUrl + file.relativePath;

        // Validate that URL starts with expected HuggingFace prefix
        const std::string expectedPrefix = "https://huggingface.co/";
        if (url.rfind(expectedPrefix, 0) != 0) {
            spdlog::error("Rejecting URL with unexpected scheme/domain: {}", url);
            allOk = false;
            continue;
        }

        // Use just the filename (last component) as the local name
        fs::path localName = fs::path(file.relativePath).filename();
        fs::path localPath = targetDir / localName;

        // Skip if file already exists with correct size
        if (fs::exists(localPath, ec) && file.sizeBytes > 0) {
            auto existingSize = fs::file_size(localPath, ec);
            if (!ec && existingSize == file.sizeBytes) {
                spdlog::info("  {} already exists, skipping",
                              localName.string());
                continue;
            }
        }

        if (!downloadFile(url, localPath)) {
            spdlog::error("  Failed to download {}", file.relativePath);
            allOk = false;
        } else {
            spdlog::info("  Downloaded {}", localName.string());
        }
    }

    if (allOk) {
        // Find the .onnx file to show the --model path
        std::string onnxFile;
        for (const auto& file : voice.files) {
            fs::path fn = fs::path(file.relativePath).filename();
            if (fn.extension() == ".onnx") {
                onnxFile = (targetDir / fn).string();
                break;
            }
        }

        std::cerr << std::endl;
        std::cerr << "Model downloaded successfully!" << std::endl;
        if (!onnxFile.empty()) {
            std::cerr << "Use with:  --model " << onnxFile << std::endl;
        }
        std::cerr << std::endl;
    }

    return allOk;
}

std::optional<fs::path> resolveModelPath(
    const std::string& nameOrAlias,
    const fs::path& modelDir) {
    auto maybeVoice = findVoice(nameOrAlias);
    if (!maybeVoice) {
        return std::nullopt;
    }

    const VoiceInfo& voice = maybeVoice.value();

    // Find the .onnx file in the voice's file list
    for (const auto& file : voice.files) {
        fs::path fn = fs::path(file.relativePath).filename();
        if (fn.extension() == ".onnx") {
            fs::path candidate = modelDir / fn;
            if (fs::exists(candidate)) {
                return candidate;
            }
        }
    }

    return std::nullopt;
}

} // namespace piper

File: src/python/piper_train/model_manager.py
"""Model manager for piper-plus voice catalog.

Provides model listing, downloading, and resolution functionality
matching the C++ model_manager.cpp and C# ModelManager.cs implementations.
"""

import os
import sys


# Embedded voice catalog (matches src/cpp/piper_plus_voices.json)
_BUILTIN_CATALOG = {
    "ja_JP-tsukuyomi-chan-medium": {
        "key": "ja_JP-tsukuyomi-chan-medium",
        "name": "tsukuyomi-chan",
        "language_code": "ja_JP",
        "language_family": "ja",
        "language_name_native": "\u65e5\u672c\u8a9e",
        "language_name_english": "Japanese",
        "quality": "medium",
        "num_speakers": 1,
        "source": "piper-plus",
        "repo_id": "ayousanz/piper-plus-tsukuyomi-chan",
        "files": {
            "tsukuyomi-chan-6lang-fp16.onnx": {"size_bytes": 39216913},
            "config.json": {"size_bytes": 8568},
        },
        "aliases": ["tsukuyomi", "tsukuyomi-chan", "ja-tsukuyomi"],
        "description": "Tsukuyomi-chan 6-language TTS model fine-tuned from multilingual base (FP16)",
    },
    "ja_JP-css10-6lang-medium": {
        "key": "ja_JP-css10-6lang-medium",
        "name": "css10-6lang",
        "language_code": "ja_JP",
        "language_family": "ja",
        "language_name_native": "\u65e5\u672c\u8a9e",
        "language_name_english": "Japanese",
        "quality": "medium",
        "num_speakers": 1,
        "source": "piper-plus",
        "repo_id": "ayousanz/piper-plus-css10-ja-6lang",
        "files": {
            "css10-ja-6lang-fp16.onnx": {"size_bytes": 39414515},
            "config.json": {"size_bytes": 8966},
        },
        "aliases": ["css10", "css10-6lang", "css10-ja", "ja-css10"],
        "description": "CSS10 Japanese 6-language TTS model fine-tuned from multilingual base (FP16, 6841 utterances)",
    },
}


def get_default_model_dir() -> str:
    """Return the default model directory (OS-specific).

    Checks PIPER_MODEL_DIR env var first, then falls back to:
    - Windows: %APPDATA%/piper/models
    - macOS: ~/Library/Application Support/piper/models
    - Linux: $XDG_DATA_HOME/piper/models or ~/.local/share/piper/models
    """
    env_dir = os.environ.get("PIPER_MODEL_DIR")
    if env_dir:
        return env_dir

    if sys.platform == "win32":
        base = os.environ.get("APPDATA", os.path.expanduser("~"))
        return os.path.join(base, "piper", "models")
    elif sys.platform == "darwin":
        return os.path.join(
            os.path.expanduser("~"), "Library", "Application Support", "piper", "models"
        )
    else:
        xdg = os.environ.get("XDG_DATA_HOME")
        if xdg:
            return os.path.join(xdg, "piper", "models")
        return os.path.join(
            os.path.expanduser("~"), ".local", "share", "piper", "models"
        )


def find_voice(name: str) -> dict | None:
    """Find a voice by key, name, or alias.

    Returns the voice dict or None if not found.
    """
    if not name:
        return None

    # Exact key match
    if name in _BUILTIN_CATALOG:
        return _BUILTIN_CATALOG[name]

    # Search by name or alias
    for voice in _BUILTIN_CATALOG.values():
        if voice["name"] == name:
            return voice
        if name in voice.get("aliases", []):
            return voice

    return None


def list_models(language_filter: str | None = None) -> None:
    """Print available voice models to stderr.

    Args:
        language_filter: Optional language code (e.g., "ja" or "ja_JP") to filter by.
    """
    voices = list(_BUILTIN_CATALOG.values())

    if language_filter:
        voices = [
            v
            for v in voices
            if v["language_family"] == language_filter
            or v["language_code"] == language_filter
        ]

    if not voices:
        if language_filter:
            print(
                f"No voice models found for language: {language_filter}",
                file=sys.stderr,
            )
        else:
            print("No voice models found.", file=sys.stderr)
        return

    print("\nAvailable voice models:", file=sys.stderr)

    current_lang = ""
    for voice in sorted(voices, key=lambda v: (v["language_code"], v["key"])):
        if voice["language_code"] != current_lang:
            current_lang = voice["language_code"]
            header = f"  {voice['language_name_english']}"
            if (
                voice.get("language_name_native")
                and voice["language_name_native"] != voice["language_name_english"]
            ):
                header += f" ({voice['language_name_native']})"
            header += f" [{current_lang}]:"
            print(f"\n{header}", file=sys.stderr)

        speakers = (
            "1 speaker"
            if voice["num_speakers"] == 1
            else f"{voice['num_speakers']} speakers"
        )
        print(
            f"    {voice['key']:<44} [{voice['source']}]  {speakers}   {voice['quality']}",
            file=sys.stderr,
        )

    print("\nUse --download-model <name> to download a model.", file=sys.stderr)


def download_model(model_name: str, model_dir: str | None = None) -> bool:
    """Download a model by name/alias to the model directory.

    Returns True on success, False on failure.
    """
    voice = find_voice(model_name)
    if voice is None:
        print(
            f"Error: Model '{model_name}' not found. Use --list-models to see available models.",
            file=sys.stderr,
        )
        return False

    if model_dir is None:
        model_dir = get_default_model_dir()

    os.makedirs(model_dir, exist_ok=True)

    print(f"Downloading model: {voice['key']} ({voice['source']})", file=sys.stderr)

    repo_id = voice["repo_id"]

    try:
        from huggingface_hub import hf_hub_download  # noqa: PLC0415
    except ImportError:
        print(
            "Error: huggingface_hub is required for model download. "
            "Install with: pip install huggingface-hub",
            file=sys.stderr,
        )
        return False

    success = True
    for filename, file_info in voice["files"].items():
        dest_path = os.path.join(model_dir, filename)

        # Skip if file already exists with correct size
        if os.path.exists(dest_path):
            existing_size = os.path.getsize(dest_path)
            expected_size = file_info.get("size_bytes", 0)
            if expected_size > 0 and existing_size == expected_size:
                print(
                    f"  Skipping {filename} (already exists, size matches)",
                    file=sys.stderr,
                )
                continue

        try:
            print(f"  Downloading {filename}...", file=sys.stderr)
            hf_hub_download(
                repo_id=repo_id,
                filename=filename,
                local_dir=model_dir,
                local_dir_use_symlinks=False,
            )
            print(f"  Saved to {dest_path}", file=sys.stderr)
        except Exception as e:
            print(f"  Failed to download {filename}: {e}", file=sys.stderr)
            success = False

    if success:
        print(f"\nModel downloaded to: {model_dir}", file=sys.stderr)

        # Print usage hint
        onnx_files = [f for f in voice["files"] if f.endswith(".onnx")]
        if onnx_files:
            onnx_path = os.path.join(model_dir, onnx_files[0])
            print("\nUsage:", file=sys.stderr)
            print(
                f"  python -m piper_train.infer_onnx --model {onnx_path} "
                f"--text '\u3053\u3093\u306b\u3061\u306f' --output-dir ./output",
                file=sys.stderr,
            )

    return success


def resolve_model_path(model_arg: str, model_dir: str | None = None) -> str | None:
    """Resolve a model argument to an actual .onnx file path.

    If model_arg is a file path that exists, return it directly.
    If it's a model name/alias, look for the downloaded model in model_dir.

    Returns the resolved path or None if not found.
    """
    # Direct file path
    if os.path.exists(model_arg):
        return model_arg

    # Try as model name/alias
    voice = find_voice(model_arg)
    if voice is None:
        return None

    if model_dir is None:
        model_dir = get_default_model_dir()

    # Find the ONNX file
    for filename in voice["files"]:
        if filename.endswith(".onnx"):
            path = os.path.join(model_dir, filename)
            if os.path.exists(path):
                return path

    return None
